presto-spi/src/main/java/io/prestosql/spi/type/UnscaledDecimal128Arithmetic.java

```diff
@@ -630,23 +630,19 @@ public static void multiply(long left, long right, Slice result)
         long z2 = 0;
         long z3 = 0;
 
-        if (l0 != 0) {
```

Comment:
I guess this `if` supposed to be an optimization. Do you say that it is opposite?

Reply:
Every `if` in the code enables the branch prediction mechanism (an probably many more) in CPU. Sometimes it is better to make few more ops than introduce branch programming. This appears to be the case since the code inside the brackets is fairly small.
And this is not the case in the generalized multiply method.
At least that what benchmarks tell me.

Reply:
How is this being measured? What benchmarks are you trying? When I wrote it this way it was precisely because some benchmarks indicated it was faster this way.

Reply:
BenchmarkDecimalOperators.
This has been discussed in https://prestosql.slack.com/archives/CP1MUNEUX/p1597832153058600.
The generic 128x128 bit method is faster with ifs, as well as 128x64 one (which is removed in this PR).
The 64x64 one is in turn faster without any branches.

Reply:
Any chance for vectorization in 64x64 in java 11?

Reply:
TLDR - no

This is actually an extremely interesting subject. I discussed it with Karol heavily.
Java will (with a little bit of help) use SIMD registers in certain situations. Unfortunately it requires operating on arrays, not single values like we do here. Making operators and functions use entire columns will likely help.
So we only need to reimplement half of core presto to make it work. Might seems a little excessive at this point though.
This - https://openjdk.java.net/jeps/338 - may help.